src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/MethodCodeNode.cs

```diff
@@ -83,9 +83,37 @@ protected override DependencyList ComputeNonRelocationBasedDependencies(NodeFact
 
             CodeBasedDependencyAlgorithm.AddDependenciesDueToMethodCodePresence(ref dependencies, factory, _method);
 
+            if (_method.IsPInvoke)
+            {
+                MethodSignature methodSig = _method.Signature;
+                if (methodSig.ReturnType.IsDelegate)
+                {
+                    AddPInvokeDelegateParameterDependencies(ref dependencies, factory, methodSig.ReturnType);
+                }
+
+                for (int i=0; i < methodSig.Length; i++)
+                {
+                    if (methodSig[i].IsDelegate)
+                    {
+                        AddPInvokeDelegateParameterDependencies(ref dependencies, factory, methodSig[i]);
+                    }
+                }
+            }
+
             return dependencies;
         }
 
+        private void AddPInvokeDelegateParameterDependencies(ref DependencyList dependencies, NodeFactory factory, TypeDesc parameter)
+        {
+            if (dependencies == null)
+                dependencies = new DependencyList();
+
+            dependencies.Add(factory.ConstructedTypeSymbol(parameter), "Delegate Marshalling Stub");
```

Comment:
I did wrote a test case  locally to verify the case you mention and this change certainly fixes it. I will include the test case here.